Document multi-tenancy with operator (#3034)
grafana/tempo: docs/sources/tempo/setup/operator/multitenancy.md

---
title: Enable multi-tenancy
description: Enable multi-tenancy
menuTitle: Enable multi-tenancy
weight: 250
aliases:
- /docs/tempo/operator/multitenancy
---

# Enable multi-tenancy

Tempo is a multi-tenant distributed tracing backend. It supports multi-tenancy through the use of a header: `X-Scope-OrgID`.
Refer to [multi-tenancy docs]({{< relref "../../operations/multitenancy" >}}) for more details. 
This document outlines how to deploy and use multi-tenant Tempo with the Operator.

## Multi-tenancy without authentication

The following Kubernetes Custom Resource (CR) deploys a multi-tenant Tempo instance.

> Note that Jaeger query is not tenant aware and therefore is not supported in this configuration.

Reviewer: Changing the note to use the admonition shortcode. 
```suggestion
{{% admonition type="note" %}}
Jaeger query is not tenant aware and therefore is not supported in this configuration.
{{% /admonition %}}
```

```yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
spec:
  tenants: {}
  storage:
    secret:
      name: minio-test
      type: s3
  storageSize: 1Gi
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
```

## OIDC authentication with static RBAC

On Kubernetes, a multi-tenant Tempo instance uses OIDC authentication and static RBAC authorization defined in the CR.
The instance should be accessed through service `tempo-simplest-gateway` which handles authentication and authorization.
The service exposes Jaeger query API and OpenTelemetry gRPC (OTLP) for trace ingestion.

> The Jaeger UI can be accessed at `http://<exposed gateway service>:8080/api/traces/v1/<tenant-name>/search`

Reviewer: ```suggestion
The Jaeger UI can be accessed at `http://<exposed gateway service>:8080/api/traces/v1/<tenant-name>/search`.
```

```yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
spec:
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
    gateway:
      enabled: true
  storage:
    secret:
      type: s3
      name: minio-test
  storageSize: 200M
  tenants:
    mode: static
    authentication:
      - tenantName: test-oidc
        tenantId: test-oidc
        oidc:
          issuerURL: http://dex.default.svc.cluster.local:30556/dex
          redirectURL: http://tempo-simplest-gateway.default.svc.cluster.local:8080/oidc/test-oidc/callback
          usernameClaim: email
          secret:
            name: oidc-test
    authorization:
      roleBindings:
      - name: "test"
        roles:
        - read-write
        subjects:
        - kind: user
          name: "admin@example.com"
      roles:
      - name: read-write
        permissions:
        - read
        - write
        resources:
        - traces
        tenants:
        - test-oidc
```

* The secret `oidc-test` defines fields `clientID`, `clientSecret` and `issuerCAPath`.

> * The RABC gives tenant `test-oidc` read and write access for traces. 

Reviewer: Should this be RBAC?

```suggestion
* The RBAC gives tenant `test-oidc` read and write access for traces. 
```

## OpenShift

> On OpenShift, the authentication and authorization does not require any third party service dependencies.

Reviewer: ```suggestion
On OpenShift, the authentication and authorization does not require any third-party service dependencies.
```

The authentication uses OpenShift OAuth (the user is redirected to the OpenShift login page) and authorization is handled through `SubjectAccessReview` (SAR).

> The instance should be accessed through service `tempo-simplest-gateway` which handles authentication and authorization.

Reviewer: ```suggestion
The instance should be accessed through service `tempo-simplest-gateway`, which handles authentication and authorization.
```

> The instance should be accessed through service `tempo-simplest-gateway` which handles authentication and authorization.
> The service exposes Jaeger query API and OpenTelemetry gRPC (OTLP) for trace ingestion.
> The Jaeger UI can be accessed at `http://<exposed gateway service>:8080/api/traces/v1/<tenant-name>/search`

Reviewer: ```suggestion
The instance should be accessed through service `tempo-simplest-gateway`, which handles authentication and authorization.
The service exposes Jaeger query API and OpenTelemetry gRPC (OTLP) for trace ingestion.
The Jaeger UI can be accessed at `http://<exposed gateway service>:8080/api/traces/v1/<tenant-name>/search`.
```

---
title: Setup multi-tenancy
description: Set up multi-tenancy
menuTitle: Multi-tenancy
weight: 250
aliases:
- /docs/tempo/operator/multitenancy
---

# Multi-tenancy

> Tempo is a multi-tenant distributed tracing backend. It supports multi-tenancy through the use of a header: `X-Scope-OrgID`.
> See [multi-tenancy docs]({{< relref "../../operations/multitenancy" >}}) for more details. This document outlines how to deploy and use multi-tenant Tempo with the Operator.

Reviewer 2: Prefer "refer to X" over "see X".
 
 nit:  [Semantic line breaks](https://sembr.org/) are better for future line based diffing in the GitHub UI and result in the same HTML output. 
 
 ```suggestion
Tempo is a multi-tenant distributed tracing backend.
It supports multi-tenancy through the use of a header: `X-Scope-OrgID`.
For more information, refer to [Enable multitenancy]({{< relref "../../operations/multitenancy" >}}). 
This document outlines how to deploy and use multi-tenant Tempo with the Operator.
```

> ## Multitenancy without authentication

Reviewer 2: Is it "multi-tenancy" or "multitenancy".

Perhaps @knylander-grafana can advise and we can lock this down in the Writers' Toolkit word list https://grafana.com/docs/writers-toolkit/write/style-guide/word-list/

Author: multi-tenancy is used in other pages

Reviewer 2: ```suggestion
## Multi-tenancy without authentication
```

> The following example CR deploys multi-tenant Tempo instance. Note that Jaeger query is not tenant aware and therefore is not supported in this configuration.

Reviewer 2: I think it makes sense to explain that abbreviation

```suggestion
The following example Kubernetes Custom Resource (CR) deploys a multi-tenant Tempo instance.
Note that Jaeger query is not tenant aware and therefore is not supported in this configuration.
```

```yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
spec:
  tenants: {}
  storage:
    secret:
      name: minio-test
      type: s3
  storageSize: 1Gi
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
```

## OIDC authentication with static RBAC

> On Kubernetes the multi-tenant Tempo instance uses OIDC authentication and static RABC authorization defined in the CR.

Reviewer 2: ```suggestion
On Kubernetes, a multi-tenant Tempo instance uses OIDC authentication and static RBAC authorization defined in the CR.
```

```yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
spec:
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
    gateway:
      enabled: true
  storage:
    secret:
      type: s3
      name: minio-test
  storageSize: 200M
  tenants:
    mode: static
    authentication:
      - tenantName: test-oidc
        tenantId: test-oidc
        oidc:
          issuerURL: http://dex.default.svc.cluster.local:30556/dex
          redirectURL: http://tempo-simplest-gateway.default.svc.cluster.local:8080/oidc/test-oidc/callback
          usernameClaim: email
          secret:
            name: oidc-test
    authorization:
      roleBindings:
      - name: "test"
        roles:
        - read-write
        subjects:
        - kind: user
          name: "admin@example.com"
      roles:
      - name: read-write
        permissions:
        - read
        - write
        resources:
        - traces
        tenants:
        - test-oidc
```

* The secret `oidc-test` defines fields `clientID`, `clientSecret` and `issuerCAPath`.
* The RABC gives tenant `test-oidc` read and write access for traces. 

## OpenShift

> On OpenShift the authentication and authorization does not require any 3rd party service dependencies.

Reviewer 2: ```suggestion
On OpenShift, the authentication and authorization does not require any third party service dependencies.
```

> The authentication uses OpenShift OAuth (the user is redirected to the OpenShift login page) and 
> authorization is handled through `SubjectAccessReview` (SAR).

Reviewer 2: nit: 
[Semantic line breaks](https://sembr.org/) are better for future line based diffing in the GitHub UI and result in the same HTML output.
```suggestion
The authentication uses OpenShift OAuth (the user is redirected to the OpenShift login page) and authorization is handled through `SubjectAccessReview` (SAR).
```

The instance should be accessed through service `tempo-simplest-gateway` which handles authentication and authorization.
The service exposes Jaeger query API and OpenTelemetry gRPC (OTLP) for trace ingestion.
The Jaeger UI can be accessed at `http://<exposed gateway service>:8080/api/traces/v1/<tenant-name>/search`

```yaml
apiVersion: tempo.grafana.com/v1alpha1
kind:  TempoStack
metadata:
  name: simplest
spec:
  storage:
    secret:
      name: object-storage
      type: s3
  storageSize: 1Gi
  tenants:
    mode: openshift
    authentication:
      - tenantName: dev
        tenantId: "1610b0c3-c509-4592-a256-a1871353dbfa"
      - tenantName: prod
        tenantId: "1610b0c3-c509-4592-a256-a1871353dbfb"
  template:
    gateway:
      enabled: true
    queryFrontend:
      jaegerQuery:
        enabled: true
```

> User is responsible for creating `ClusterRole` and `ClusterRoleBinding` for reading and writing the data.

Reviewer 2: Who is the user here? Can we explain this a little more?
